core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java

```diff
@@ -0,0 +1,132 @@
+ * desired for the index creation operation.
+ */
+public abstract class IndexCreationMonitor<Response extends CreateIndexResponse> extends AbstractComponent {
+
+    private final ClusterService clusterService;
+    private final ThreadContext threadContext;
+
+    protected IndexCreationMonitor(Settings settings, ClusterService clusterService, ThreadContext threadContext) {
+        super(settings);
+        this.clusterService = Objects.requireNonNull(clusterService);
+        this.threadContext = Objects.requireNonNull(threadContext);
+    }
+
+    public void waitOnShards(final String indexName,
+                             final WriteConsistencyLevel writeConsistencyLevel,
+                             final ActionListener<Response> listener,
+                             final TimeValue timeout) {
+
+        final ClusterStateObserver observer = new ClusterStateObserver(clusterService, logger, threadContext);
+        final ClusterState observedState = observer.observedState();
+        final ClusterStateObserver.ChangePredicate shardsAllocatedPredicate = new ClusterStateObserver.ChangePredicate() {
+            @Override
+            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                return newStatus == ClusterState.ClusterStateStatus.APPLIED
+                           && enoughShardsActive(newState, indexName, writeConsistencyLevel);
+            }
+
+            @Override
+            public boolean apply(ClusterChangedEvent changedEvent) {
+                return enoughShardsActive(changedEvent.state(), indexName, writeConsistencyLevel);
+            }
+        };
+        final ClusterStateObserver.Listener observerListener = new ClusterStateObserver.Listener() {
+            @Override
+            public void onNewClusterState(ClusterState state) {
+                listener.onResponse(newResponse(true));
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                logger.warn("[{}] cluster service closed while waiting for write consistency number of shards " +
+                                "to be allocated on index creation.", indexName);
+                listener.onResponse(newResponse(false));
+            }
+
+            @Override
+            public void onTimeout(TimeValue timeout) {
+                logger.warn("[{}] index was created, but the operation timed out waiting for write consistency " +
+                                "number of shards to be allocated.", indexName);
+                listener.onResponse(newResponse(false));
+            }
+        };
+
+        if (observedState.status() == ClusterState.ClusterStateStatus.APPLIED
+                && enoughShardsActive(observedState, indexName, writeConsistencyLevel)) {
+            observerListener.onNewClusterState(observedState);
```

Comment:
Maybe this check should be above the listener/predicate construction? Sort of a "quick return" style?